snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -0,0 +1,150 @@
+from collections import deque
+from datetime import datetime
+from enum import Enum
+from typing import Deque, Mapping, Optional, cast
+
+from arroyo import Message
+from arroyo.processing.strategies import ProcessingStrategy
+
+from snuba.subscriptions.utils import Tick
+from snuba.utils.metrics import MetricsBackend
+
+
+class SchedulerMode(Enum):
+    IMMEDIATE = "immediate"
+    WAIT_FOR_SLOWEST_PARTITION = "wait-for-slowest"
+
+
+class TickBuffer(ProcessingStrategy[Tick]):
+    """
+    The TickBuffer buffers ticks until they are ready to be submitted to
+    the next processing step.
+
+    The behavior of the TickBuffer depends on which of the two scheduler
+    modes applies.
+
+    If the scheduler mode is IMMEDIATE then there is no buffering and a
+    message is always immediately submitted to the next processing step.
+
+    If the scheduler mode is WAIT_FOR_SLOWEST_PARTITION then messages are
+    buffered until all partitions are at least up to the timestamp of the
+    tick.
+
+    `max_ticks_buffered_per_partition` applies if the scheduler mode is
+    WAIT_FOR_SLOWEST_PARTITION. Once the maximum ticks is received for that
+    partition, we start to submit ticks for processing even if that timestamp
+    is not received for all partitions yet.
+
+    This prevents the buffered tick list growing infinitely and scheduling
+    to grind to a halt if one partition starts falling far behind for some reason.
+    """
+
+    def __init__(
+        self,
+        mode: SchedulerMode,
+        partitions: int,
+        max_ticks_buffered_per_partition: Optional[int],
+        next_step: ProcessingStrategy[Tick],
+        metrics: MetricsBackend,
+    ) -> None:
+        if mode == SchedulerMode.WAIT_FOR_SLOWEST_PARTITION:
+            assert max_ticks_buffered_per_partition is not None
```

Comment:
I think making this optional gives the false impression that is is ok to infinitely buffer ticks when it is not. The problem if this starts to error is that even on a restart it will continue to error and won't be able to make any progress. If we have an incident where a single partition falls far behind the others (i.e. we hit some buffer limit and start erroring) we would not be able to recover without manual intervention probably from someone who has at least some context about what this code is doing. I'd worry that it can compound our problems in the middle of an incident. Now it may be that we consider this scenario unlikely and don't want to worry about it, but this was my reason for forcing a value to be provided.

We can add other type of alerting on this later that does not actually error so hitting the buffer max still happens loudly enough for us to know something is wrong.

Reply:
What if we had a topic where consistency across partition was so important that we did not want to allow it to proceed and where the manual intervention was preferrable. 
It should not be the common practice as most topics would not need it, but still allowed. 
I am ok adding it separately, though please still add warnings logged when we decide to move ahead anyway.

Reply:
Yeah if that requirement ever comes up, it can be added as a 3rd strategy